### PR TITLE
skip compression on upgrade connection

### LINKED ISSUE #182: [bug] Compression middleware should step out of the way on upgrade requests
See https://github.com/gorilla/websocket/issues/589 for a complete description of the problem.  

The summary is that compression middleware writes to hijacked connections [on this line](https://github.com/gorilla/handlers/blob/f08afc1876ad882db8074bcb8a4cc96107d3a5f4/compress.go#L112) and the HTTP server panics as a result.

One fix for the issue is to detect hijack and avoid writing to the connection after hijack.  

The larger problem is that the compression middleware should not do anything on a protocol upgrade request.  The middleware assumes HTTP, but the connection will be used for a different protocol.

The middleware should call the handler directly when the connection upgrade token is present. Add the following to the beginning of the handler:

    if httpguts.HeaderValuesContainsToken(r.Header["Connection"], "Upgrade") {
        h.ServeHTTP(w, r)
        return
    }

If you don't want to add a dependency on [httpguts](https://godoc.org/golang.org/x/net/http/httpguts), then do something like this:

    var upgrade bool
    for _, v := range strings.Split(r.Header.Get("Connection"), ",") {
        if strings.ToLower(strings.TrimSpace(v)) == "upgrade" {
            upgrade = true
            break
        }
    }
    if upgrade {
        h.ServeHTTP(w, r)
        return
    }

### PR DESCRIPTION
**Pull request by benitogf**
Fixes #182 

**Summary of Changes**

1. Followed what was mentioned on the issue

### CODE REVIEW

**Comment by benitogf** [comment]
bump

File: compress.go [hunk 1]
```diff
@@ -102,6 +102,18 @@ func CompressHandlerLevel(h http.Handler, level int) http.Handler {
...
+		for _, v := range strings.Split(r.Header.Get("Connection"), ",") {
+			if strings.ToLower(strings.TrimSpace(v)) == "upgrade" {
+				upgrade = true
+				break
+			}
+		}
+		if upgrade {
+			h.ServeHTTP(w, r)
+			return
+		}
```
Comment (blixt) [inline compress.go:115]: This should be sufficient:

```suggestion
		if r.Header.Get("Upgrade") != "" {
			h.ServeHTTP(w, r)
			return
		}
```

File: compress.go [hunk 2]
```diff
@@ -102,6 +102,11 @@ func CompressHandlerLevel(h http.Handler, level int) http.Handler {
 			return
 		}
 
+		if r.Header.Get("Upgrade") != "" {
```
Comment (elithrar) [inline compress.go:105]: See the review comment.

Comment (blixt) [reply compress.go:105]: Which comment?

Comment (elithrar) [reply compress.go:105]: https://github.com/gorilla/handlers/pull/187#discussion_r466991374v - maybe I misunderstood here.

Comment (blixt) [reply compress.go:105]: Ah! The proposal I made was already pushed, these 4 lines checking the `Upgrade` header should work fine (I've been using a wrapper implementing exactly this in production on framer.com for many months and haven't run into any issues with WebSockets.)

**Pull request merged**